### ISSUE TITLE
Buffering a selected polygon doesn't seem to work

### ISSUE THREAD

**Issue by tchaddad** [geomoose/gm3#365]
As mentioned in https://github.com/geomoose/gm3/issues/48, selecting a polygon, then applying a buffer, in order to select neighboring properties, is a common workflow.

Attempting to do this in the demo, here is what I do:

- Step 1: Select polygon with point
- Step 2: Return to select menu, and switch to "using Feature from Results", and "with buffer 300 feet"
- Step 3: press Go

The result is that the the buffer is applied to my point, and not to the selected polygon, and so the number of selected parcels is incorrect. In fact, as shown in the screenshots below, for a parcel with dimensions larger than 300 feet, it is possible that the buffer lies entirely inside the parcel, and so no neighbors are selected.

Is my method incorrect, or is buffer working on the wrong object?

_____________

**Step 1**
![image](https://user-images.githubusercontent.com/1042618/53617803-a2475980-3b9d-11e9-8b41-155f36b320d8.png)

_____________

**Step 2**
![image](https://user-images.githubusercontent.com/1042618/53617828-be4afb00-3b9d-11e9-8477-56976dbe3910.png)

_____________

**Step 3**
![image](https://user-images.githubusercontent.com/1042618/53617838-d6227f00-3b9d-11e9-8816-21032e42381c.png)

**Comment by klassenjs** [comment]
Looks like you didn't click on the parcel to select it (after choosing select from results).  The parcel shouldn't be yellow.

**Comment by tchaddad** [comment]
Ok - thanks, that's helpful, although not necessarily intuitive. How does the user know that they need to select something again that looks like it is already selected?

Not arguing, just making a case for a better narrative so that users don't make this mistake.

Where would I look to edit the instructions that show up in the select panel? I'll see if I can come up with language that can be put into a pull request...

**Comment by klassenjs** [comment]
One thing that might be a bug is it "remembering" what was previously selected (the point from the previous select).  If it didn't do that, it would give you an error telling you that you need to select something.

Here is where to start looking: https://github.com/geomoose/gm3/blob/master/src/services/select.js

**Comment by klassenjs** [comment]
Although that remembering can also be really useful if you want to re-run your last query with different options.

**Comment by chughes-lincoln** [comment]
Is there any way currently to buffer multiple results? This appears to only let you do a search or a selection, and then buffer one of the results. It would be useful to be able to buffer multiple parcels (ie in a certain subdivision or zoning) rather than one parcel.

**Comment by theduckylittle** [comment]
It appears to be working, but yes, I can see where there is some confusion as the old point is still visible on the map. There is some debate to be had about whether to clear the original point (losing a previous polygon can be a loss, if for example, one wants to edit it afterwards).

@chughes-lincoln  - We can put that in the request queue but we've sometimes decided to table certain features to prevent complicating the interface any more than it already is...

**Comment by theduckylittle** [comment]
Not sure how to best tag this... do we need a doc???

**Comment by chughes-lincoln** [comment]
Personally I think the main issue is that it isn't very intuitive. There's no way for a user to know that they have to click on the map to select a previous result to buffer it rather than just choosing feature from results and clicking Go to buffer the selection. I had one of our GIS Analysts try to do it and he needed my help to explain how the tool should be used, which would be even more difficult for a basic user who isn't as accustomed to GIS.

Not sure what the solution is. Perhaps modifying the labels in the select tab to emphasize that you have to click on the map to buffer a result or to emphasize that whatever selection geometry you see on the map is what will be used for the selection.

**Comment by theduckylittle** [comment]
Keeping this open as a discussion point for 4.0.

**Comment by tchaddad** [comment]
Looking at this again, in light of the change made for (https://github.com/geomoose/gm3/issues/397), I think the intuitive flow of select is better now, but 

the issue now is that the "Using" list includes up to 8 options:
- 6 options work on initial use of select tool (Box, Point, Multipoint, Line, Polygon, Drawings/Markup)
- 2 options only work on a selection that already exists (Results, Modify Feature)

Only thought that comes to mind is that if it were feasible to somehow visually break the "Using" list into 2 parts to separate out the options that require a selection to already exist, that would probably be a further usability improvement for newcomers. 

If a physical break isn't feasible, perhaps a style indicator (like bold, italics or another font color) could be used to indicate that those options have a precondition that must be satisfied before they are used?

**Comment by chughes-lincoln** [comment]
Failing making the buffer more intuitive, it would also be useful to be add an html link to a pdf explaining how to use the select/search tools and the buffer. I tried messing around with the select.js but I couldn't figure out an easy way like there is in geomoose 2. This wouldn't solve all the issues because it assumes people would read it and understand it, but it would at least give people direct guidance.

**Comment by chughes-lincoln** [comment]
Related to this, after doing an initial selection, if you select the feature from results option without clicking on a result (thinking it will buffer all results), and click GO after adding a buffer distance, it will buffer your initial selection geometry which isn't what the user expects. It would be nice if it would throw a warning that you need to select a feature from results if that option is selected.

**Comment by chughes-lincoln** [comment]
Also, would separating the buffer from the selection into it's own tool be a solution (if that's even possible)? i.e. not changing how the buffer works at all, just moving the parameters into a separate tab. I think people intuitively understand you need results to use a buffer, but if they're in the same tool people will think you'll be able to do both at the same time.

**Comment by klassenjs** [comment]
You don't need results to use the buffer.

You can buffer any vector feature be it one you just drew or one selected from a vector layer, including from the scratch layer (where you can draw or load/save features to GeoJSON/KML) and from the results layer from a previous query.

**Comment by chughes-lincoln** [comment]
Would it be more intuitive to have a separate buffer for the results though? I think the main issue is that people don't understand that the buffer in the initial selection is buffering the input, not the results you'd get from the input. 

If you moved the 'Feature from Results/Drawing' to it's own tool (even if that's more of a cosmetic change rather than creating a new tool) I think that would clear up some of the confusion. You could even keep the 'With buffer' option in the select tool to keep the ability to buffer your input (perhaps changing the language to 'With buffer around input' so no one things this is buffering results?). 

I don't know all that would go into changing this, but I think as is, the average user wouldn't understand how to properly use the buffer and that it's a greater issue that just changing the language of it.

**Comment by klassenjs** [comment]
I agree people do generally need to be told about the ability to chain queries and that they can use any vector source as an input though before they get it.

What I don't want to loose is the power and flexibility for people to explore the data that we get from this chaining workflow that would be lost in a more directed and intuitive tool.  

That's not to say that adding a tool to answer an especially common question more intuitively/quickly is a bad thing.  But it is then likely useful only for that specific case.

**Comment by elil** [comment]
@klassenjs This may be a case for configuration options.  This tool can really do a lot!  The longer this thread goes, the more powerful (and complex) it becomes.  

An underlying tool with these capabilities is really useful and not limited in anyway.  With configuration options, this same tool could be exposed in a more simple way, "a tool to answer an especially common question more intuitively/quickly... useful only for that specific case," which sounds perfect for some users.  The same tool could even be exposed two or three times as different services with different configurations so that there are a few distinct, simple tools that do just what you want.

**Comment by chughes-lincoln** [comment]
Regarding setting up the buffer to be able to buffer multiple results, I agree that we don't want to make the buffer more complicated than it already is. Would it be feasible to set it up so that holding Ctrl while clicking on the map allows you to add multiple features? I don't know how difficult this would be so I'm just making the suggestion, but that wouldn't complicate the tool at all.

**Comment by klassenjs** [comment]
I don't know about feasibility, but I fully agree that would be nice to operate on multiple features from the result set.
